lib/posix-netlink/netlink.c

```diff
@@ -0,0 +1,309 @@
+		goto err_out;
+	}
+
+	nl_ctx = uk_zalloc(drv->allocator, sizeof(*nl_ctx));
+	if (unlikely(!nl_ctx)) {
+		err = ENOMEM;
+		goto err_out;
+	}
+	nl_ctx->a = drv->allocator;
+	nl_ctx->drv = nl_drv;
+#if CONFIG_LIBPOSIX_PROCESS
+	/* FIXME: PID is here like port number and must be unique. It just
+	 *        starts with PID if free.
+	 */
+	nl_ctx->nl_pid = getpid();
+#endif /* CONFIG_LIBPOSIX_PROCESS */
+
+	nl_ctx->nl_recvqueue = uk_mbox_create(nl_ctx->a, 512);
+	if (unlikely(!nl_ctx->nl_recvqueue)) {
+		err = ENOMEM;
+		goto err_free_ctx;
+	}
+
+	/* Optional driver initialization */
+	if (nl_ctx->drv->ops->create) {
+		err = nl_ctx->drv->ops->create(nl_ctx);
+		if (unlikely(err))
+			goto err_free_recvqueue;
+	}
+
+	uk_pr_debug("Created netlink socket: %p, protocol: %d, drv: %s\n",
+		    nl_ctx, nl_ctx->drv->protocol, nl_ctx->drv->libname);
+	return nl_ctx;
+
+err_free_recvqueue:
+	uk_mbox_free(nl_ctx->a, nl_ctx->nl_recvqueue);
+err_free_ctx:
+	uk_free(nl_ctx->a, nl_ctx);
+err_out:
+	return ERR2PTR(-err);
+}
+
+static int nl_close(posix_sock *s)
+{
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	struct uk_streambuf *nlbuf;
+
+	if (nl_ctx->drv->ops->close)
+		nl_ctx->drv->ops->close(nl_ctx);
+
+	/* Release queued and unsent messages */
+	while (uk_mbox_recv_try(nl_ctx->nl_recvqueue, (void **)&nlbuf) == 0) {
+		uk_pr_debug("Releasing unconsumed netlink message %p\n", nlbuf);
+		nlbuf_free(nlbuf);
+	}
+
+	uk_mbox_free(nl_ctx->a, nl_ctx->nl_recvqueue);
+	uk_free(nl_ctx->a, nl_ctx);
+	return 0;
+}
+
+static int nl_bind(posix_sock *s,
+		   const struct sockaddr *addr,
+		   socklen_t addr_len)
+{
+	const struct sockaddr_nl *nl_addr;
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+
+	if (unlikely(nl_ctx->bound))
+		return -EADDRINUSE;
+	if (unlikely(!addr || addr_len < sizeof(*nl_addr)))
+		return -EINVAL;
+	nl_addr = (const struct sockaddr_nl *)addr;
+
+	if (unlikely((nl_addr->nl_family != AF_NETLINK) ||
+		     (nl_addr->nl_pad != 0))
+		return -EINVAL;
+
+	/* We only support the client to connect to us: the kernel */
+	if (nl_addr->nl_pid != 0)
+		return -EHOSTUNREACH;
+
+	nl_ctx->nl_groups = nl_addr->nl_groups;
+	nl_ctx->bound = 1;
+	return 0;
+}
+
+static int nl_sockname(posix_sock *s,
+		       struct sockaddr *restrict addr,
+		       socklen_t *restrict addr_len)
+{
+	struct sockaddr_nl *nl_addr;
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+
+	if (*addr_len < sizeof(*nl_addr)) {
+		/* addr_len is too small, only inform about needed size */
+		goto out;
+	}
+
+	nl_addr = (struct sockaddr_nl *)addr;
+	nl_addr->nl_family = AF_NETLINK;
+	nl_addr->nl_pid = nl_ctx_pid(nl_ctx);
+	nl_addr->nl_pad = 0;
+	nl_addr->nl_groups = nl_ctx->nl_groups;
+
+out:
+	*addr_len = sizeof(*nl_addr);
+	return 0;
+}
+
+static inline int nl_handle(struct nl_ctx *nl_ctx, const void *buf, size_t len)
+{
+	const struct nlmsghdr *nlh;
+	size_t tmp_len;
+	int ret = 0;
+
+	UK_ASSERT(nl_ctx);
+	UK_ASSERT(nl_ctx->drv);
+	UK_ASSERT(nl_ctx->drv->ops);
+	UK_ASSERT(nl_ctx->drv->ops->handle);
```

Comment:
so why is only `handle` asserted for but `close`/`create` are actively checked for? maybe this should also be documented that only `handle` is mandatory?